spec/unit/pops/validator/validator_spec.rb

```diff
@@ -34,6 +34,56 @@ def validate(factory)
+  end
+
+  it 'should not raise error for legal definition locations' do
+    expect(validate(parse('function aaa::bbb() {}',      'aaa/manifests/bbb.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+    expect(validate(parse('define bbb() {}',      'aaa/manifests/site.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+    expect(validate(parse('class aaa() {}',      'aaa/manifests/init.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+    expect(validate(parse('class aaa() {}',      'manifests/site.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+    expect(validate(parse('function aaa::bbb::ccc() {}', 'aaa/manifests/bBb.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+    expect(validate(parse('function aaa::bbb::ccc() {}', 'aaa/manifests/bbb/CCC.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+  end
+
+  it 'should not raise error for class locations when not parsing a file' do
+    #nil file means eval or some other way to get puppet language source code into the catalog
+    expect(validate(parse('function aaa::ccc() {}', nil))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+  end
+
+  it 'should not raise error for definitions inside a manifest file' do
+    Puppet[:manifest] = 'a/manifest/file.pp'
+    expect(validate(parse('class aaa() {}', 'a/manifest/file.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+  end
+
+  it 'should not raise error for definitions inside a manifest directory' do
+    Puppet[:manifest] = 'a/manifest/dir'
+    expect(validate(parse('class aaa() {}', 'a/manifest/dir/file1.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+    expect(validate(parse('class bbb::ccc::ddd() {}', 'a/manifest/dir/and/more/stuff.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+  end
+
+  it 'should raise error for definitions not inside a manifest directory' do
+    Puppet[:manifest] = 'a/manifest/somewhere/else'
+    expect(validate(parse('class aaa() {}', 'a/manifest/dir/file1.pp'))).to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+  end
+
+  # Regexp duration was 0.055935
+  # String.start_with? duration was 0.057642
+  # context 'microbenchmark' do
+  #   start = Time.now
+  #   (1..1000).each do
+  #     it 'Should have or not have errors at the right time...' do
+  #       expect(validate(parse('function aaa::bbb() {}',      'aaa/manifests/bbb.pp'))).not_to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+  #       expect(validate(parse('define aaa::bbb::ccc::eee() {}', 'aaa/manifests/bbb/ddd.pp'))).to have_issue(Puppet::Pops::Issues::ILLEGAL_DEFINITION_LOCATION)
+  #     end
+  #   end
+  #   duration = Time.now - start
+  #   puts "duration was %{duration}" % { duration: duration }
+  # end
```

Comment:
The regexp vs start_with is dwarfed by everything else. The micro benchmark I used shows regexp to take 3x as long. And the measurement in the benchmark above is too short running so the diff is within the margin of error.